### PR TITLE
Add Get-StringHash and Get-FileHash cmdlets

### LINKED ISSUE #2731: Rewrite Get-FileHash on C#
As approved in #2718  Get-FileHash should be rewritten on C#.

### PR DESCRIPTION
**Pull request by iSazonov**
Close #2731 

1. Rewrite Get-FileHash on C# and refactoring tests
The original Get-FileHash output type is `Microsoft.Powershell.Utility.FileHash`. The type was defined by ExtendedTypeDefinition. It is not C# type and reported as `PSCustomObject`.
Therefore, the output type was changed and now it is native C# type `Microsoft.PowerShell.Commands.FileHashInfo`.

2.  Add new cmdlet Get-StringHash and add tests
The original idea was to add a InputString parameter to the common cmdlet but due to a conflict with the Path parameter (paths is strings too!) I had to make a new cmdlet to provide a pipeline support for the InputString parameter.

### CODE REVIEW

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FileHash.Tests.ps1
Comment (lzybkr) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FileHash.Tests.ps1]: This test should not be necessary, `PSDefaultParameterValues` is already well tested.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FileHash.Tests.ps1
Comment (lzybkr) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FileHash.Tests.ps1]: Should be moved closer to the use to limit the scope of the variable.

**Review by lzybkr — CHANGES_REQUESTED** [review]
@PowerShell/powershell-committee - needs to review the new cmdlet - parameters, output type, allowing null.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-StringHash.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-StringHash.Tests.ps1]: We can change it after the merger.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-StringHash.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-StringHash.Tests.ps1]: Fixed.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-StringHash.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-StringHash.Tests.ps1]: Fixed.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-StringHash.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-StringHash.Tests.ps1]: It is from original tests.
Removed.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-StringHash.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-StringHash.Tests.ps1]: I thought about this example:
```powershell
$stringarray | Get-StringHash 
```
If the string array contains a null string member then we get throw - is this the desired behavior?
The following example does not cause a throw:
```powershell
"a",$null | Select-String -Pattern ".*"
```

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs]: Get-Uptime was merged and #2659 was opened for URI.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs]: Oh, sorry. I remember your earlier comment about this, but all mixed up :-)
Fixed.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs]: Fixed.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs]: It is from original cmdlet and `HashAlgorithmNames` class.
If there are other useful algorithms, please give a link.
And whether we should do something specifically for Mac?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs]: If we block it by `ValidateSet`, an user will not be able to type an algorithm name in low cases - is it okay?

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FileHash.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FileHash.Tests.ps1]: Fixed.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FileHash.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FileHash.Tests.ps1]: Fixed.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FileHash.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FileHash.Tests.ps1]: Removed.

File: src/System.Management.Automation/commands/utility/FormatAndOutput/common/DefaultFormatters/PowerShellCore_format_ps1xml.cs
Comment (iSazonov) [inline src/System.Management.Automation/commands/utility/FormatAndOutput/common/DefaultFormatters/PowerShellCore_format_ps1xml.cs]: Fixed.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs]: What I found:
1. `Unblock-File` use the same pattern.
2. For `Path` parameter `GetResolvedProviderPathFromPSPath` has the same problem - "the file might disappear between this check and when we try to read it."

So we could keep consistency here.

Later we try read the file and _ignore_ any exceptions.
The original `Get-FileHash` function call [Write-Error and return](https://github.com/PowerShell/PowerShell/blob/master/src/Modules/Shared/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psm1#L128)
Should we reproduce the same behavior or  _ignore_ any exceptions?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs
Comment (lzybkr) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs]: Never mind - `ValidateSet` ignores case.

Maybe not applicable here, but it's often useful to store the exact argument a user specifies so error messages or whatever look like what they wrote. Of course that means you couldn't use `switch` as easily, but I don't think that's a big deal.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs
Comment (lzybkr) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs]: According to the `ValidateSet` use [here](https://github.com/PowerShell/PowerShell/blob/e568fda5c3dfd85ea18c0406217533849bb6e966/src/Modules/Shared/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psm1#L18), the original function has:

* MACTripleDES
* MD5
* RIPEMD160
* SHA1
* SHA256
* SHA384
* SHA512

I suppose it's not surprising that a couple of these are missing for .Net Core.

We should probably add the missing algorithms for Windows PowerShell or explore alternative implementations to that we can have completely portable scripts.

/cc @SteveL-MSFT

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs
Comment (lzybkr) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs]: Unsound patterns should not be copied, even for consistency.

We should introduce sound patterns, and when feasible and time allows, correct the unsound patterns.

Comment (lzybkr) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs]: Also - ignoring exceptions is usually a terrible idea - I think non-terminating errors is definitely preferred.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-StringHash.Tests.ps1
Comment (lzybkr) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-StringHash.Tests.ps1]: `$null` is not passed through the pipeline and `Select-String` allows `$null` anyway, so that's not a good example.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FileHash.Tests.ps1
Comment (lzybkr) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FileHash.Tests.ps1]: Can't this just use `$testDocument = Join-Path ...`?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs
Comment (SteveL-MSFT) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs]: @lzybkr if .Net Core doesn't support those, we should treat it as a separate issue (not block this PR).  however, we should see if .Net Core will implement them or why they no longer support it (patents?  no longer secure?)

Comment (lzybkr) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs]: I believe these are hashing algorithms, so security isn't the issue. My guess is less demand.

But the real question here - are we OK with a PR that would be a breaking change to Windows PowerShell if we took this code back to Windows as is?

Comment (SteveL-MSFT) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs]: The security concern is that hashes are used to determine if something (like a file/package) has been modified.  However, SHA1, for example, has been proven that you can generate collisions meaning that it is feasible someone could modify a file and produce the same resulting SHA1 hash so it appears to be legit.

We can discuss this at the Committee meeting.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-StringHash.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-StringHash.Tests.ps1]: Yes, but key is that `null` string don't throw an exception here.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs]: Fixed.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FileHash.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FileHash.Tests.ps1]: Fixed.

**Comment by iSazonov** [comment]
@PowerShell/powershell-committeeTeam 
Should we introduce a new type 'Microsoft.PowerShell.Commands.StringHashInfo' for `Get-StringHash` output similar to `Get-FileHash` output?

**Comment by lzybkr** [comment]
@PowerShell/powershell-committee has decided we need an RFC for the `Get-StringHash` cmdlet.

The biggest concern was how PowerShell will convert anything to `string`, so the cmdlet will happily hash objects in a potentially unexpected way.

The output type was also of some concern when combined with allowing `null` because it will be difficult to match input with output. We think the output needs the original object in addition to the hash.

**Comment by lzybkr** [comment]
The committee also agreed that we do not need any updates with regards to the algorithms supported - we've marked the PR wit `Breaking Change` so we can document the difference between Windows PowerShell and PowerShell Core.

**Comment by iSazonov** [comment]
What will be the road map for the PR?
Could I remove `Get-StringHash` code from the PR?

I do not have the forces for the RFC to write the big text correctly in English. 
Can we ask the author @rkeithhill of the `Get-StringHash` idea to start this RFC?

**Comment by lzybkr** [comment]
By all means - remove `Get-StringHash` from the PR.

Your English seems more than sufficient to write a RFC, but if @rkeithhill is up for it, I assume he'll jump in.

**Comment by iSazonov** [comment]
@lzybkr I create new PR with RFC https://github.com/PowerShell/PowerShell-RFC/pull/63

Now `Get-StringHash` was removed from the PR.
Travis again temporary failed 😕

**Comment by iSazonov** [comment]
@lzybkr  Should I do something else in the PR?

**Review by lzybkr — APPROVED** [review]

**Comment by iSazonov** [comment]
@lzybkr Thanks!

**Pull request merged**
